### PR TITLE
2.7 support. travis integration testing. API Added

### PR DESCRIPTION
**Pull request by Madoshakalaka**
Hi,

I loved your sysdm so much and I happen to need an api in my project that I stayed up late writing a pull request.

There are these changes

- blackened project files
- python 2.7 better surpported. There were previous compatibility issues from syntax error that prevents sysdm from starting under 2.7, e.g. encoding. And there were issue like non-existent FileNotExistError in python 2.7
- added [dev] extra dependency with pytest and tox
- introduced travis script that does integration testing on 2.7 3.4 3.5 3.6 3.7 (tox dropped support for 3.3, so I gave 3.3 up)
See tests here:
https://travis-ci.org/Madoshakalaka/sysdm/builds/594527527
- Added a simple API that lists enabled units
- polished readme help
- refactored some function names to make them more descriptive and added docstrings

### CODE REVIEW

File: deploy.py [hunk 1]
```diff
@@ -22,7 +24,9 @@
     init = f.read()
 
 with open("sysdm/__init__.py", "w") as f:
-    f.write(re.sub('__version__ = "[0-9.]+"', '__version__ = "{}"'.format(version), init))
+    f.write(
+        re.sub('__version__ = "[0-9.]+"', '__version__ = "{}"'.format(version), init)
+    )
 
 os.system("rm -rf dist/")
 os.system("python setup.py sdist bdist_wheel")
```
Comment (Madoshakalaka) [inline deploy.py]: Formatting change by black

File: setup.py [hunk 2]
```diff
@@ -10,41 +10,43 @@
...
+        "Topic :: Software Development :: Libraries",
+        "Topic :: Software Development :: Libraries :: Python Modules",
+        "Topic :: System :: Software Distribution",
+        "Topic :: System :: Systems Administration",
+        "Topic :: Utilities",
     ],
     zip_safe=False,
-    platforms='posix',
+    platforms="posix",
 )
```
Comment (Madoshakalaka) [inline setup.py:52]: updated supported version. black formmatting.

File: sysdm/__init__.py [hunk 3]
```diff
@@ -16,3 +20,22 @@ def version():
...
+    :return: Dict[str, Tuple[bool, bool, str] ]  a dict in the form of {unit_name:  (running, enabled, port)}.
+        port can be an empty string
+    """
+    if systempath is None:
+        systempath = get_default_systempath()
+    systempath = os.path.expanduser(systempath)
+    systempath = systempath.rstrip("/")
+
+    unit_names = get_created_unit_names(systempath)
+    return get_unit_info_by_names(unit_names, systempath)
```
Comment (Madoshakalaka) [inline sysdm/__init__.py:41]: Added API that simply lists units created by sysdm

File: sysdm/__main__.py [hunk 4]
```diff
@@ -258,6 +273,6 @@ def _main():
 
 def main():
     try:
-        _main()
+        _main(argv=sys.argv)
     except KeyboardInterrupt:
         print("Aborted")
```
Comment (Madoshakalaka) [inline sysdm/__main__.py:259]: argv=sys.argv for testability

**Comment by Madoshakalaka** [comment]
I used an environment variable TRAVIS to let the integration tests run only on travis, worring running sysdm with tox multiple times messes up development environment. 

It may actually not be a big deal. As long as created sysdm services in each tests get deleted at the end of the test.

**Comment by Madoshakalaka** [comment]
The integration test is a crude starting point.

It has some problems with `systemctl` on travis. I believe systemctl is not supported on travis.org

**Comment by kootenpv** [comment]
Hey thanks a lot! I will make sure to merge in all your efforts.

Note I am already using `black`, however, I have the `--line-length` set to `100` (the only configurable param in black, and I simply disagree with 88).

Also, I am curious, did you get it working in Travis?

I will have a closer look at "listing units" to make sure we get it correctly (the logic for sudo/normal user was quite nasty and took me some time to get right, I hope you did too :) )

**Comment by kootenpv** [comment]
Maybe we could have `list_unit_info` return classes that allow inspecting the attributes `is_enabled`, `is_running`, `port`, and allow functions to be called on them like `enable`, `disable`, `start`, `stop`, or you prefer something that can be serialized? Perhaps JSON makes sense then.

**Comment by Madoshakalaka** [comment]
I'm writing a daemon with your sysdm with a cli interface. And I need cli to be able to know whether it's running or not.

when the daemon is running
```
$ my_program do_task
doing task...
```

when the daemon is not running
```
$ my_program do_task
starting program in the background...
doing task
```

Either of your ideas work, serialization is not required though.

Additionaly I want to be able to start/stop/set as startup program from my cli.

I'm being a bit lazy here and didn't write API for those, as those are easily achievable with subprocess call to sysdm. I did the minimum effort to add the `list_units` functionality as it's not easy to achieve.

And yes, the tests are working

https://travis-ci.org/Madoshakalaka/sysdm/builds/594527527
Screenshot from 2019-10-07 16-43-59

Just with the caveate that with the systemctl problem, scripts are only created but not running. The tests only test for the consistency of the API.
 
You can see the log of this specific failed build during my revision:
https://travis-ci.org/Madoshakalaka/sysdm/jobs/594524419

```
Failed to connect to bus: No such file or directory
```
is printed to stderr when systemctl is called from python

**Comment by Madoshakalaka** [comment]
Oh...seems like it's not easy to enable/disable, run/stop scripts (at least not all of them). I can expand the API and add them later and if I do I'll make sure to use a `line-length` of 100

**Comment by Madoshakalaka** [comment]
@kootenpv I'll close this for now. I haven't been working on this for a long time.

Thanks for your timely support! Sysdm is amazing :+1:

**Comment by kootenpv** [comment]
I am so sad I missed all your further work on this :(

I hope you still like sysdm @Madoshakalaka!

**Pull request closed without merging**